### PR TITLE
Add Background 3D Preview export to File menu.

### PR DESCRIPTION
**Pull request by ikeough**
### Purpose

This PR adds Background 3D Preview export to the File menu. It enables the uses to export the current Background 3D Preview, with the superimposed grid as a .png file at screen resolution. A future PR will add the ability to specify the resolution of the image. It re-uses the same menu icon for image capture, as seen in the image below.

image

image

### Declarations

Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [x] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [x] Snapshot of UI changes, if any.

### Reviewers

@Racel Is this the appropriate location in the UI for this functionality?
@mjkkirschner You and I talked about the image export code at Smart Geometry.

### CODE REVIEW

**Comment by Racel** [comment]
@ikeough - this looks good to me and seems to be in the appropriate place in application menu. A few things:
- when a user switches to background preview, will the camera button in the top right of Dynamo switch to capture the background preview? I think it should and the tooltip should also be updated.
- can you please add ellipses to after "Export Background 3D Preview as Image" to be consistent with the other menu items?

**Comment by kronz** [comment]
@ikeough I appreciate the thought, but why not as a node?  The thing I hear about this functionalty is being to automate generating design state recording or creating animations.  STL export is a little weird here too, and might not be a good model.  Didn't we talk at some point about making it an optional port on watch 3d (which can get a little more specific about what parts of the graphics to export?

**Comment by ikeough** [comment]
@kronz I made it as a node at Smart Geometry. We can have that too. This functionality is important for simply being able to export the entirety of the 3D image at a printable resolution. STL export is not weird here. It follows the pattern many 3D applications where saving or exporting to different file formats of the entire model is offered in the File menu.

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs [hunk 1]
```diff
@@ -1824,6 +1824,12 @@ public void SaveImage(object parameters)
                 "NodeCount", CurrentSpace.Nodes.Count());
         }
 
+        public void Save3DImage(object parameters)
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs]: if we're adding this to the public API should document it especially what goes in this parameter

File: src/DynamoCoreWpf/Views/Core/DynamoView.xaml [hunk 2]
```diff
@@ -365,6 +365,17 @@
                             </MenuItem.Icon>
                         </MenuItem>
                         <MenuItem Focusable="False"
+                                  Header="{x:Static p:Resources.DynamoViewFileMenuExport3DAsImage}"
+                                  Command="{Binding ShowSaveImageDialogAndSaveResultCommand}"
+                                  CommandParameter="3D"
```
Comment (mjkkirschner) [inline src/DynamoCoreWpf/Views/Core/DynamoView.xaml]: can we pull this out somewhere as a constant?

**Comment by mjkkirschner** [comment]
@ikeough possible to add a test for this functionality as well?

**Comment by mjkkirschner** [comment]
@ikeough @kronz I agree that both of these feel kind of strange... I'd argue .stl export is weirder than this. What does it export, all visible geometry? All geometry? 

Is Dynamo a 3d modeling program?

**Comment by ikeough** [comment]
@mjkkirschner @kronz The name says it all "Export Background 3D Preview as Image."

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs (hunk 1 above)
Comment (ikeough) [inline src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs]: Eeeek. Sorry Mike. This does not need to be public.

**Comment by Racel** [comment]
@ikeough @mjkkirschner @kronz - I don't think the export image options are weird since they are about taking a snapshot of the views you are working with. I do,however, question the STL export as this one is about capturing the 'thing' that you made using Dynamo. The STL export option feels more like it should be a node. Just my two cents.

I also consider animating geometry slightly different than the snapshots we create with the 2 export image options. While we may not actually be transforming the geometry (from what I understand), I think users would think that they were transforming geometry or the camera (at least temporarily) for any animation/key frame type nodes.

**Comment by andydandy74** [comment]
If I may chime in - I think it would be nice to have both functionalities (3D background image export and STL export) available in the File menu as well as in the form of nodes. Having it in the file menu is great for casual exporting ("quickly taking a snapshot") while having it as nodes is great for more developed workflows.

**Comment by mjkkirschner** [comment]
@ikeough everything looks good except the string parameters... can we pull them out to a private enum or some constants defined somewhere? then LGTM

**Comment by ikeough** [comment]
@mjkkirschner `internal const` wouldn't work for the command parameters because the values have to be public to be used in the xaml. I've added them to our resource files instead. This could pass as an API change, because the resource files generate a public property with `{get;set;}`.

**Comment by mjkkirschner** [comment]
@ikeough thanks for the changes, I forget one other issue, not sure if a test is possible (easy?) for this functionality before merge. Leave it up to you.

**Comment by ikeough** [comment]
@mjkkirschner Yeah, trying to figure out how that would work currently. It uses the same method internally for saving image files as was used previously by the command,, with a bit of conditional logic for whether the command parameter exists and has certain values. It's hard to test this because both paths produce images.

**Comment by mjkkirschner** [comment]
could check the color value of a specific pixel or hash it in some simple way, like sum of colors?

**Comment by ikeough** [comment]
This should probably be using two different commands. Then the logic would be clearer and the testing would be more straightforward. But, I didn't want to add a new public command.

**Comment by mjkkirschner** [comment]
@kronz any reason to *not* merge this before the branch cut?

also @ikeough just curious, did you get rid of the auto property because of an old mono limitation?

**Comment by ikeough** [comment]
@mjkkirschner Yeah. Auto-properties seem to break the build system.

**Comment by kronz** [comment]
@mjkkirschner @ikeough Sorry, I was out the last 2 days.  The branch has been cut now, Ian, not sure if there is some particular customer facing event you wanted to get this in for?  I'm fine with it just going itn 1.3.

**Comment by ikeough** [comment]
@kronz I'll bring it it now for 1.3. It's for putting together images for publication.

**Comment by alfarok** [comment]
@ikeough I see this functionality in the daily builds but not in 1.3.0, has it been pushed back to possibly 1.4?  Just curious because I have been working on a package that would greatly benefit from its integration.

[HydraShare](http://hydrashare.github.io/hydra/viewer?owner=alfarok&fork=hydra&id=dynamo_test&slide=0&scale=1&offset=0,0)

**Comment by Racel** [comment]
@alfarok - unfortunately, it looks like this was not cherry-picked into 1.3. It should be in the next release.

**Comment by alfarok** [comment]
@Racel no problem, thanks for the update!

**Pull request merged**